### PR TITLE
[Core] Zero-copy asdict for InputMetadata

### PR DESCRIPTION
**Pull request by Yard1**
<details>

<summary><b> PR Checklist (Click to expand. Please read before submitting.) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

---

`dataclasses.asdict` will deepcopy tensors inside `InputMetadata`, leading to an avoidable overhead. This PR adds a new method to `InputMetadata` which will return a dict of fields and their values without copy.

Before:
Screenshot 2024-03-18 at 1 30 04 PM

After:
Screenshot 2024-03-18 at 1 29 56 PM

### CODE REVIEW

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -527,7 +526,7 @@ def prepare_input_tensors(
                 "lora_requests": lora_requests,
                 "lora_mapping": lora_mapping,
             }
-            metadata_dict.update(dataclasses.asdict(input_metadata))
+            metadata_dict.update(input_metadata.asdict_zerocopy())
```
Comment (simon-mo) [inline vllm/worker/model_runner.py:529]: can you just do `input_metadata.__dict__` to avoid the new method at all?

Comment (Yard1) [reply vllm/worker/model_runner.py:529]: Perhaps, though that may lead to unexpected interactions if `InputMetadata` has extra properties added and such.

**Comment by youkaichao** [comment]
It seems the deepcopy of `dataclasses.asdict` is notorious, see https://github.com/python/cpython/issues/88071 . 

How about moving to `attrs`? https://www.attrs.org/en/stable/index.html it is not a builtin package, but just a third-party library.

```python
from attrs import asdict, define
@define
class MyDataClass:
    a: int
    b: list
instance = MyDataClass(1, [2, 3])
data1 = asdict(instance, recurse=False)
assert data1['b'] is instance.b
```

Its `asdict` has a `recursive` parameter.

**Comment by youkaichao** [comment]
Here is the whole picture:

```python
import attrs
@attrs.define(slots=False)
class MyDataClass:
    a: int
    b: list
    def __attrs_post_init__(self):
        self.c = None
instance = MyDataClass(1, [2, 3])
data1 = asdict(instance, recurse=False)
assert data1['b'] is instance.b # make sure it is shallow copy
assert "c" not in data1 # make sure asdict only contains annotated fields
```

**Comment by Yard1** [comment]
I don't think it makes sense to add a third-party library just for this one specific usecase.

**Comment by youkaichao** [comment]
Oh, `attrs` is quite a famous library, and it motivates the builtin `dataclasses`.

One reason that `attrs` does not become builtin, is because `attrs` does not want to be bound with python versions. So you can basically think `attrs` as an advanced replacement of `dataclasses`.

e.g. when you want to use new features from `dataclasses`, you have to upgrade `python` interpreter; however, with `attrs`, you just need to update a package.

I vote for `attrs` because later I plan to refactor `vllm/config.py` to use `dataclass`-like features (so that we can easily print/log them for debugging). And I might switch to `attrs` in all the places.

Open to any feedback :)

**Comment by Yard1** [comment]
I think it's fine if your PR makes use of it, but I don't think it's necessary to couple `attrs` with this particular PR :)

**Comment by youkaichao** [comment]
Then please make sure to rename `asdict_zerocopy` to `_asdict_zerocopy`, so that users know this is private and they won't call it.

**Comment by WoosukKwon** [comment]
> Then please make sure to rename asdict_zerocopy to _asdict_zerocopy, so that users know this is private and they won't call it.

I think it should be `asdict_zerocopy` since it's outside `InputMetadata`?

File: vllm/model_executor/input_metadata.py [hunk 2]
```diff
@@ -31,3 +31,9 @@ class InputMetadata:
     def __post_init__(self):
         # will not appear in the __repr__ and __init__
         self.attn_bias = None
+
+    def asdict_zerocopy(self) -> dict:
```
Comment (WoosukKwon) [inline vllm/model_executor/input_metadata.py:35]: nit:
```suggestion
    def asdict_zerocopy(self) -> Dict[str, Any]:
```

Comment (WoosukKwon) [reply vllm/model_executor/input_metadata.py]: You should also add `from typing import Any, Dict` 😅

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for finding the performance bug. Didn't know the issue in `dataclasses.asdict`. TIL 😄

**Comment by WoosukKwon** [comment]
@Yard1 It seems the CI failed. Could you please double check?

**Comment by Yard1** [comment]
yeah let me see

**Comment by Yard1** [comment]
Should be good now!

**Pull request merged**